src/env/double_write.rs

```diff
@@ -0,0 +1,722 @@
+        }
+    }
+}
+
+impl<F: FileSystem> Drop for HedgedHandle<F> {
+    fn drop(&mut self) {
+        self.disk1.send((FileTask::Stop, Box::new(|_| {}))).unwrap();
+        self.disk2.send((FileTask::Stop, Box::new(|_| {}))).unwrap();
+        self.handle1.take().unwrap().join().unwrap();
+        self.handle2.take().unwrap().join().unwrap();
+    }
+}
+
+impl<F: FileSystem> Handle for HedgedHandle<F> {
+    fn truncate(&self, offset: usize) -> IoResult<()> {
+        block_on(self.wait_one(FileTask::Truncate(offset))).map(|_| ())
+    }
+
+    fn file_size(&self) -> IoResult<usize> {
+        block_on(self.wait_one(FileTask::FileSize)).map(|x| x.unwrap())
+    }
+
+    fn sync(&self) -> IoResult<()> {
+        block_on(self.wait_one(FileTask::Sync)).map(|_| ())
+    }
+}
+
+pub struct HedgedWriter<F: FileSystem> {
+    disk1: Sender<(FileTask, Callback<usize>)>,
+    disk2: Sender<(FileTask, Callback<usize>)>,
+    writer1: Arc<F::Writer>,
+    writer2: Arc<F::Writer>,
+}
+
+impl<F: FileSystem> HedgedWriter<F> {
+    pub fn new(disk1: Sender<(FileTask, Callback<usize>)>, disk2: Sender<(FileTask, Callback<usize>)>, writer1: Arc<F::Writer>, writer2: Arc<F::Writer>) -> Self {
+      
+        Self {
+            handle,
+
+        }
+    }
+}
+
+impl<F: FileSystem> Write for HedgedWriter<F> {
+    fn write(&mut self, buf: &[u8]) -> IoResult<usize> {
+    }
+
+    fn flush(&mut self) -> IoResult<()> {
+    }
+}
+
+impl<F: FileSystem> WriteExt for HedgedWriter<F> {
+    fn truncate(&mut self, offset: usize) -> IoResult<()> {
+    }
+
+    fn allocate(&mut self, offset: usize, size: usize) -> IoResult<()> {
+    }
+}
+
+impl<F: FileSystem> Seek for HedgedWriter<F> {
+    fn seek(&mut self, pos: SeekFrom) -> IoResult<u64> {
+        match pos {
+            SeekFrom::Start(offset) => self.offset = offset as usize,
+            SeekFrom::Current(i) => self.offset = (self.offset as i64 + i) as usize,
+            SeekFrom::End(i) => self.offset = (self.inner.file_size()? as i64 + i) as usize,
+        }
+        Ok(self.offset as u64)
+    }
+}
+
+pub struct HedgedReader<F: FileSystem> {
+    inner: Arc<HedgedHandle<F>>,
+    offset: usize,
+}
+
+impl<F: FileSystem> HedgedReader<F> {
+    pub fn new(handle: Arc<HedgedHandle<F>>) -> Self {
+        Self {
+            inner: handle,
+            offset: 0,
+        }
+    }
+}
+
+impl<F: FileSystem> Seek for HedgedReader<F> {
+    fn seek(&mut self, pos: SeekFrom) -> IoResult<u64> {
+        match pos {
+            SeekFrom::Start(offset) => self.offset = offset as usize,
+            SeekFrom::Current(i) => self.offset = (self.offset as i64 + i) as usize,
+            SeekFrom::End(i) => self.offset = (self.inner.file_size()? as i64 + i) as usize,
+        }
+        Ok(self.offset as u64)
+    }
+}
+
+impl<F: FileSystem> Read for HedgedReader<F> {
+    fn read(&mut self, buf: &mut [u8]) -> IoResult<usize> {
+        let len = self.inner.read(self.offset, buf)?;
+        self.offset += len;
```

Comment:
you mentioned that there might be concurrent readers,  is it safe doing `self.offset += len` from multiple threads?

Reply:
Note that there would be multiple readers, but for one reader, it's not shared between threads. It doesn't implement `Sync` trait, so the rust compiler would make sure it is not shared between threads.

Reply:
Understand,  looks like it's a sole append-only writer + multiple readers design